libtransform/transform_util.3
.\" Copyright (c) 2003-2007 Tim Kientzle
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: head/lib/libtransform/transform_util.3 201098 2009-12-28 02:58:14Z kientzle $
.\"
.Dd January 8, 2005
.Dt transform_util 3
.Os
.Sh NAME
.Nm transform_clear_error ,
.Nm transform_compression ,
.Nm transform_compression_name ,
.Nm transform_copy_error ,
.Nm transform_errno ,
.Nm transform_error_string ,
.Nm transform_file_count ,
.Nm transform_filter_code ,
.Nm transform_filter_count ,
.Nm transform_filter_name ,
.Nm transform_format ,
.Nm transform_format_name ,
.Nm transform_position ,
.Nm transform_set_error
.Nd libtransform utility functions
.Sh SYNOPSIS
.In transform.h
.Ft void
.Fn transform_clear_error "struct transform *"
.Ft int
.Fn transform_compression "struct transform *"
.Ft const char *
.Fn transform_compression_name "struct transform *"
.Ft void
.Fn transform_copy_error "struct transform *" "struct transform *"
.Ft int
.Fn transform_errno "struct transform *"
.Ft const char *
.Fn transform_error_string "struct transform *"
.Ft int
.Fn transform_file_count "struct transform *"
.Ft int
.Fn transform_filter_code "struct transform *" "int"
.Ft int
.Fn transform_filter_count "struct transform *" "int"
.Ft const char *
.Fn transform_filter_name "struct transform *" "int"
.Ft int
.Fn transform_format "struct transform *"
.Ft const char *
.Fn transform_format_name "struct transform *"
.Ft int64_t
.Fn transform_position "struct transform *" "int"
.Ft void
.Fo transform_set_error
.Fa "struct transform *"
.Fa "int error_code"
.Fa "const char *fmt"
.Fa "..."
.Fc
.Sh DESCRIPTION
These functions provide access to various information about the
.Tn struct transform
object used in the
.Xr libtransform 3
library.
.Bl -tag -compact -width indent
.It Fn transform_clear_error
Clears any error information left over from a previous call.
Not generally used in client code.
.It Fn transform_compression
Synonym for
.Fn transform_filter_code(a, 0) .
.It Fn transform_compression_name
Synonym for
.Fn transform_filter_name(a, 0) .
.It Fn transform_copy_error
Copies error information from one transform to another.
.It Fn transform_errno
Returns a numeric error code (see
.Xr errno 2 )
indicating the reason for the most recent error return.
Note that this can not be reliably used to detect whether an
error has occurred.
It should be used only after another libtransform function
has returned an error status.
.It Fn transform_error_string
Returns a textual error message suitable for display.
The error message here is usually more specific than that
obtained from passing the result of
.Fn transform_errno
to
.Xr strerror 3 .
.It Fn transform_file_count
Returns a count of the number of files processed by this transform object.
The count is incremented by calls to
.Xr transform_write_header
or
.Xr transform_read_next_header .
.It Fn transform_filter_code
Returns a numeric code identifying the indicated filter.
See
.Fn transform_filter_count
for details of the numbering.
.It Fn transform_filter_count
Returns the number of filters in the current pipeline.
For read transform handles, these filters are added automatically
by the automatic format detection.
For write transform handles, these filters are added by calls to the various
.Fn transform_write_add_filter_XXX
functions.
Filters in the resulting pipeline are numbered so that filter 0
is the filter closest to the format handler.
As a convenience, functions that expect a filter number will
accept -1 as a synonym for the highest-numbered filter.
.Pp
For example, when reading a uuencoded gzipped tar transform, there
are three filters:
filter 0 is the gunzip filter,
filter 1 is the uudecode filter,
and filter 2 is the pseudo-filter that wraps the transform read functions.
In this case, requesting
.Fn transform_position(a, -1)
would be a synonym for
.Fn transform_position(a, 2)
which would return the number of bytes currently read from the transform, while
.Fn transform_position(a, 1)
would return the number of bytes after uudecoding, and
.Fn transform_position(a, 0)
would return the number of bytes after decompression.
.It Fn transform_filter_name
Returns a textual name identifying the indicated filter.
See
.Fn transform_filter_count
for details of the numbering.
.It Fn transform_format
Returns a numeric code indicating the format of the current
transform entry.
This value is set by a successful call to
.Fn transform_read_next_header .
Note that it is common for this value to change from
entry to entry.
For example, a tar transform might have several entries that
utilize GNU tar extensions and several entries that do not.
These entries will have different format codes.
.It Fn transform_format_name
A textual description of the format of the current entry.
.It Fn transform_position
Returns the number of bytes read from or written to the indicated filter.
In particular,
.Fn transform_position(a, 0)
returns the number of bytes read or written by the format handler, while
.Fn transform_position(a, -1)
returns the number of bytes read or written to the transform.
See
.Fn transform_filter_count
for details of the numbering here.
.It Fn transform_set_error
Sets the numeric error code and error description that will be returned
by
.Fn transform_errno
and
.Fn transform_error_string .
This function should be used within I/O callbacks to set system-specific
error codes and error descriptions.
This function accepts a printf-like format string and arguments.
However, you should be careful to use only the following printf
format specifiers:
.Dq %c ,
.Dq %d ,
.Dq %jd ,
.Dq %jo ,
.Dq %ju ,
.Dq %jx ,
.Dq %ld ,
.Dq %lo ,
.Dq %lu ,
.Dq %lx ,
.Dq %o ,
.Dq %u ,
.Dq %s ,
.Dq %x ,
.Dq %% .
Field-width specifiers and other printf features are
not uniformly supported and should not be used.
.El
.Sh SEE ALSO
.Xr transform_read 3 ,
.Xr transform_write 3 ,
.Xr libtransform 3 ,
.Xr printf 3
.Sh HISTORY
The
.Nm libtransform
library first appeared in
.Fx 5.3 .
.Sh AUTHORS
.An -nosplit
The
.Nm libtransform
library was written by
.An Tim Kientzle Aq kientzle@acm.org .


NEWS

Jan 02, 2010: libarchive 2.7.902a released (test release for 2.8)
Jan 02, 2010: Fix tar/test/test_windows on MinGW
Jan 02, 2010: Fix memory leaks in libarchive tests
Jan 01, 2010: Fix memory leak when filter startup fails

Dec 27, 2009: libarchive 2.7.901a released (test release for 2.8)

Aug 04, 2009: libarchive 2.7.1 released
Jul 20, 2009: Suppress bogus warning about unxz
Jul 19, 2009: Support Cygwin 1.7
Jun 11, 2009: Support lzma/xz files compressed with larger buffer sizes.
May 24, 2009: Handle gzip files signed with OpenBSD "gzsig" program.
May 07, 2009: Avoid false failures when reading from pipe.

Apr 16, 2009: libarchive 2.7.0 released

Apr 10, 2009: libarchive 2.6.992a released
Apr 09, 2009: Fix SIGPIPE issue building with MSVC.
Apr 09, 2009: Fix several minor memory leaks in libarchive and libarchive_test

Apr 08, 2009: libarchive 2.6.991a released
Apr 07, 2009: Additional tests added to bsdcpio_test

Apr 01, 2009: libarchive 2.6.990a released
Apr 01, 2009: Use command-line gunzip, bunzip2, unxz, unlzma for
	decompression if the library is built without suitable
	libraries.  The setup functions return ARCHIVE_WARN
	in this case so clients can adapt if necessary.
Apr 01, 2009: Use getpw*_r and getgr*_r functions for thread-safety.
Mar 24, 2009: Add transform_read_next_header2(), which is up to 25%
	more efficient for some clients; from Brian Harring.
Mar 22, 2009: PDF versions of manpages are now included in the distribution.
Mar, 2009: Major work to improve Cygwin build by Charles Wilson.
Feb/Mar, 2009: Major work on cmake build support, mostly by Michihiro NAKAJIMA.
Feb/Mar, 2009: Major work on Visual Studio support by Michihiro NAKAJIMA.
	All tests now pass.
Feb 25, 2009: Fix Debian Bug #516577
Feb 21, 2009: Yacc is no longer needed to build; date parser rewritten in C.
Jan/Feb, 2009: Mtree work by Michihiro.
Feb, 2009: Joliet support by Andreas Henriksson.
Jan/Feb, 2009: New options framework by Michihiro.
Feb, 2009: High-res timestamps on Tru64, AIX, and GNU Hurd, by Björn Jacke.
Jan 18, 2009: Extended attributes work on FreeBSD and Linux now	with pax format.
Jan 07, 2009: New transform_read_disk_entry_from_file() knows about ACLs,
	extended attributes, etc so that bsdtar and bsdcpio don't require
	such system-specific knowledge.
Jan 03, 2009: Read filter system extensively refactored.  In particular,
	read filter pipelines are now built out automatically and individual
	filters should be much easier to implement.  Documentation on the
	Googlecode Wiki explains how to implement new filters.
Dec 28, 2008: Many Windows/Visual Studio fixes from Michihiro NAKAJIMA.

Dec 28, 2008: Main libarchive development moved from FreeBSD Perforce
	server to Google Code.  This should make it easier for more
	people to participate in libarchive development.

Dec 28, 2008: libarchive 2.6.0 released
Dec 25, 2008: libarchive 2.5.905a released
Dec 10, 2008: libarchive 2.5.904a released
Dec 04, 2008: libarchive 2.5.903a released
Nov 09, 2008: libarchive 2.5.902a released
Nov 08, 2008: libarchive 2.5.901a released
Nov 08, 2008: Start of pre-release testing for libarchive 2.6

Nov 07, 2008: Read filter refactor: The decompression routines just
	consume and produce arbitrarily-sized blocks.  The reblocking
	from read_support_compression_none() has been pulled into the
	read core.  Also, the decompression bid now makes multiple
	passes and stacks read filters.
Oct 21, 2008: bsdcpio: New command-line parser.
Oct 19, 2008: Internal read_ahead change: short reads are now an error
Oct 06, 2008: bsdtar: option parser no longer uses getopt_long(),
	gives consistent option parsing on all platforms.
Sep 19, 2008: Jaakko Heinonen: shar utility built on libarchive
Sep 17, 2008: Pedro Giffuni: birthtime support
Sep 17, 2008: Miklos Vajna: lzma reader and test.  Note:  I still have
	some concerns about the auto-detection (LZMA file format
	doesn't support auto-detection well), so this is not yet
	enabled under transform_read_support_compression_all().  For
	now, you must call transform_read_support_compression_lzma() if
	you want LZMA read support.
Sep 11, 2008: Ivailo Petrov: Many fixes to Windows build, new solution files
Jul 26, 2008: archive_entry now tracks which values have not been set.
	This helps zip extraction (file size is often "unknown") and
	time restores (tar usually doesn't know atime).
Jul 26, 2008: Joerg Sonnenberger: Performance improvements to shar writer
Jul 25, 2008: Joerg Sonnenberger: mtree write support

Jul 02, 2008: libarchive 2.5.5 released

Jul 02, 2008: libarchive 2.5.5b released
Jul 01, 2008: bsdcpio is being used by enough people, we can call it 1.0.0 now
Jun 20, 2008: bsdcpio: If a -l link fails with EXDEV, copy the file instead
Jun 19, 2008: bsdcpio: additional long options for better GNU cpio compat
Jun 15, 2008: Many small portability and bugfixes since 2.5.4b.

May 25, 2008: libarchive 2.5.4b released
May 21, 2008: Joerg Sonnenberger: fix bsdtar hardlink handling for newc format

May 21, 2008: More progress on Windows building.  Thanks to "Scott"
        for the Windows makefiles, thanks to Kees Zeelenberg for
        code contributions.

May 21, 2008: Fix a number of non-exploitable integer and buffer overflows,
        thanks to David Remahl at Apple for pointing these out.

May 21, 2008: Colin Percival: SIGINFO or SIGUSR1 to bsdtar prints progress info

May 16, 2008: bsdtar's test harness no longer depends on file ordering.
        This was causing spurious test failures on a lot of systems.
        Thanks to Bernhard R. Link for the diagnosis.

May 14, 2008: Joerg Sonnenberger: -s substitution support for bsdtar

May 13, 2008: Joerg Sonnenberger: Many mtree improvements

May 11, 2008: Joerg Sonnenberger: fix hardlink extraction when
	hardlinks have different permissions from original file

April 30, 2008: Primary libarchive work has been moved into the FreeBSD
	project's Perforce repository: 	http://perforce.freebsd.org/
	The libarchive project can be browsed at

libtransform/transform_util.c
/*-
 * Copyright (c) 2009,2010 Michihiro NAKAJIMA
 * Copyright (c) 2003-2007 Tim Kientzle
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "transform_platform.h"
__FBSDID("$FreeBSD: head/lib/libtransform/transform_util.c 201098 2009-12-28 02:58:14Z kientzle $");

#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#ifdef HAVE_FCNTL_H
#include <fcntl.h>
#endif
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#ifdef HAVE_STRING_H
#include <string.h>
#endif

#include "transform.h"
#include "transform_private.h"
#include "transform_string.h"

int
transform_version_number(void)
{
	return (TRANSFORM_VERSION_NUMBER);
}

const char *
transform_version_string(void)
{
	return (TRANSFORM_VERSION_STRING);
}

int
transform_errno(struct transform *a)
{
	return (a->transform_error_number);
}

const char *
transform_error_string(struct transform *a)
{

	if (a->error != NULL  &&  *a->error != '\0')
		return (a->error);
	else
		return ("(Empty error message)");
}

int
transform_compression(struct transform *a)
{
	return transform_filter_code(a, 0);
}

const char *
transform_compression_name(struct transform *a)
{
	return transform_filter_name(a, 0);
}


/*
 * Return a count of the number of compressed bytes processed.
 */
int64_t
transform_position_compressed(struct transform *a)
{
	return transform_filter_bytes(a, -1);
}

/*
 * Return a count of the number of uncompressed bytes processed.
 */
int64_t
transform_position_uncompressed(struct transform *a)
{
	return transform_filter_bytes(a, 0);
}

void
transform_clear_error(struct transform *a)
{
	transform_string_empty(&a->error_string);
	a->error = NULL;
	a->transform_error_number = 0;
}

void
transform_set_error(struct transform *a, int error_number, const char *fmt, ...)
{
	va_list ap;

	a->transform_error_number = error_number;
	if (fmt == NULL) {
		a->error = NULL;
		return;
	}

	va_start(ap, fmt);
	transform_string_vsprintf(&(a->error_string), fmt, ap);
	va_end(ap);
	a->error = a->error_string.s;
}

void
transform_copy_error(struct transform *dest, struct transform *src)
{
	dest->transform_error_number = src->transform_error_number;

	transform_string_copy(&dest->error_string, &src->error_string);
	dest->error = dest->error_string.s;
}

void
__transform_errx(int retvalue, const char *msg)
{
	static const char *msg1 = "Fatal Internal Error in libtransform: ";
	size_t s;

	s = write(2, msg1, strlen(msg1));
	(void)s; /* UNUSED */
	s = write(2, msg, strlen(msg));
	(void)s; /* UNUSED */
	s = write(2, "\n", 1);
	(void)s; /* UNUSED */
	exit(retvalue);
}

/*
 * Parse option strings
 *  Detail of option format.
 *    - The option can accept:
 *     "opt-name", "!opt-name", "opt-name=value".
 *
 *    - The option entries are separated by comma.
 *        e.g  "compression=9,opt=XXX,opt-b=ZZZ"
 *
 *    - The name of option string consist of '-' and alphabet
 *      but character '-' cannot be used for the first character.
 *      (Regular expression is [a-z][-a-z]+)
 *
 *    - For a specfic format/filter, using the format name with ':'.
 *        e.g  "zip:compression=9"
 *        (This "compression=9" option entry is for "zip" format only)
 *
 *      If another entries follow it, those are not for
 *      the specfic format/filter.
 *        e.g  handle "zip:compression=9,opt=XXX,opt-b=ZZZ"
 *          "zip" format/filter handler will get "compression=9"
 *          all format/filter handler will get "opt=XXX"
 *          all format/filter handler will get "opt-b=ZZZ"
 *
 *    - Whitespace and tab are bypassed.
 *
 */
int
__transform_parse_options(const char *p, const char *fn, int keysize, char *key,
    int valsize, char *val)
{
	const char *p_org;
	int apply;
	int kidx, vidx;
	int negative; 
	enum {
		/* Requested for initialization. */
		INIT,
		/* Finding format/filter-name and option-name. */
		F_BOTH,
		/* Finding option-name only.
		 * (already detected format/filter-name) */
		F_NAME,
		/* Getting option-value. */
		G_VALUE,
	} state;

	p_org = p;
	state = INIT;
	kidx = vidx = negative = 0;
	apply = 1;
	while (*p) {
		switch (state) {
		case INIT:
			kidx = vidx = 0;
			negative = 0;
			apply = 1;
			state = F_BOTH;
			break;
		case F_BOTH:
		case F_NAME:
			if ((*p >= 'a' && *p <= 'z') ||
			    (*p >= '0' && *p <= '9') || *p == '-') {
				if (kidx == 0 && !(*p >= 'a' && *p <= 'z'))
					/* Illegal sequence. */
					return (-1);
				if (kidx >= keysize -1)
					/* Too many characters. */
					return (-1);
				key[kidx++] = *p++;
			} else if (*p == '!') {
				if (kidx != 0)
					/* Illegal sequence. */
					return (-1);
				negative = 1;
				++p;
			} else if (*p == ',') {
				if (kidx == 0)
					/* Illegal sequence. */
					return (-1);
				if (!negative)
					val[vidx++] = '1';
				/* We have got boolean option data. */
				++p;
				if (apply)
					goto complete;
				else
					/* This option does not apply to the
					 * format which the fn variable
					 * indicate. */
					state = INIT;
			} else if (*p == ':') {
				/* obuf data is format name */
				if (state == F_NAME)
					/* We already found it. */
					return (-1);
				if (kidx == 0)
					/* Illegal sequence. */
					return (-1);
				if (negative)
					/* We cannot accept "!format-name:". */
					return (-1);
				key[kidx] = '\0';
				if (strcmp(fn, key) != 0)
					/* This option does not apply to the
					 * format which the fn variable
					 * indicate. */
					apply = 0;
				kidx = 0;
				++p;
				state = F_NAME;
			} else if (*p == '=') {
				if (kidx == 0)
					/* Illegal sequence. */
					return (-1);
				if (negative)
					/* We cannot accept "!opt-name=value". */
					return (-1);
				++p;
				state = G_VALUE;
			} else if (*p == ' ') {
				/* Pass the space character */
				++p;
			} else {
				/* Illegal character. */
				return (-1);
			}
			break;
		case G_VALUE:
			if (*p == ',') {
				if (vidx == 0)
					/* Illegal sequence. */
					return (-1);
				/* We have got option data. */
				++p;
				if (apply)
					goto complete;
				else
					/* This option does not apply to the
					 * format which the fn variable
					 * indicate. */
					state = INIT;
			} else if (*p == ' ') {
				/* Pass the space character */
				++p;
			} else {
				if (vidx >= valsize -1)
					/* Too many characters. */
					return (-1);
				val[vidx++] = *p++;
			}
			break;
		} 
	}

	switch (state) {
	case F_BOTH:
	case F_NAME:
		if (kidx != 0) {
			if (!negative)
				val[vidx++] = '1';
			/* We have got boolean option. */
			if (apply)
				/* This option apply to the format which the
				 * fn variable indicate. */
				goto complete;
		}
		break;
	case G_VALUE:
		if (vidx == 0)
			/* Illegal sequence. */
			return (-1);
		/* We have got option value. */
		if (apply)
			/* This option apply to the format which the fn
			 * variable indicate. */
			goto complete;
		break;
	case INIT:/* nothing */
		break;
	}

	/* End of Option string. */
	return (0);

complete:
	key[kidx] = '\0';
	val[vidx] = '\0';
	/* Return a size which we've consumed for detecting option */
	return ((int)(p - p_org));
}

/*
 * Create a temporary file
 */
#if !defined(_WIN32) || defined(__CYGWIN__)

static int
get_tempdir(struct transform_string *temppath)
{
	const char *tmp;

	tmp = getenv("TMPDIR");
	if (tmp == NULL)
#ifdef _PATH_TMP
		tmp = _PATH_TMP;
#else
                tmp = "/tmp";
#endif
	transform_strcpy(temppath, tmp);
	if (temppath->s[temppath->length-1] != '/')
		transform_strappend_char(temppath, '/');
	return (TRANSFORM_OK);
}

#if defined(HAVE_MKSTEMP)

/*
 * We can use mkstemp().
 */

int
__transform_mktemp(const char *tmpdir)
{
	struct transform_string temp_name;
	int fd;

	transform_string_init(&temp_name);
	if (tmpdir == NULL) {
		if (get_tempdir(&temp_name) != TRANSFORM_OK)
			goto exit_tmpfile;
	} else {
		transform_strcpy(&temp_name, tmpdir);
		if (temp_name.s[temp_name.length-1] != '/')
			transform_strappend_char(&temp_name, '/');
	}
	transform_strcat(&temp_name, "libtransform_XXXXXX");
	fd = mkstemp(temp_name.s);
	if (fd < 0)
		goto exit_tmpfile;
	unlink(temp_name.s);
exit_tmpfile:
	transform_string_free(&temp_name);
	return (fd);
}

#else

/*
 * We use a private routine.
 */

int
__transform_mktemp(const char *tmpdir)
{
        static const char num[] = {
		'0', '1', '2', '3', '4', '5', '6', '7',
		'8', '9', 'A', 'B', 'C', 'D', 'E', 'F',
		'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N',
		'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V',
		'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd',
		'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l',
		'm', 'n', 'o', 'p', 'q', 'r', 's', 't',
		'u', 'v', 'w', 'x', 'y', 'z'
        };
	struct transform_string temp_name;
	struct stat st;
	int fd;
	char *tp, *ep;
	unsigned seed;

	fd = -1;
	transform_string_init(&temp_name);
	if (tmpdir == NULL) {
		if (get_tempdir(&temp_name) != TRANSFORM_OK)
			goto exit_tmpfile;
	} else
		transform_strcpy(&temp_name, tmpdir);
	if (temp_name.s[temp_name.length-1] == '/') {
		temp_name.s[temp_name.length-1] = '\0';
		temp_name.length --;
	}
	if (stat(temp_name.s, &st) < 0)
		goto exit_tmpfile;
	if (!S_ISDIR(st.st_mode)) {
		errno = ENOTDIR;
		goto exit_tmpfile;
	}
	transform_strcat(&temp_name, "/libtransform_");
	tp = temp_name.s + transform_strlen(&temp_name);
	transform_strcat(&temp_name, "XXXXXXXXXX");
	ep = temp_name.s + transform_strlen(&temp_name);

	fd = open("/dev/random", O_RDONLY);
	if (fd < 0)
		seed = time(NULL);
	else {
		if (read(fd, &seed, sizeof(seed)) < 0)
			seed = time(NULL);
		close(fd);
	}
	do {
		char *p;

		p = tp;
		while (p < ep)
			*p++ = num[((unsigned)rand_r(&seed)) % sizeof(num)];
		fd = open(temp_name.s, O_CREAT | O_EXCL | O_RDWR, 0600);
	} while (fd < 0 && errno == EEXIST);
	if (fd < 0)
		goto exit_tmpfile;
	unlink(temp_name.s);
exit_tmpfile:
	transform_string_free(&temp_name);
	return (fd);
}

#endif /* HAVE_MKSTEMP */
#endif /* !_WIN32 || __CYGWIN__ */
